app/services/admin/api/cms/template_service.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+class Admin::Api::CMS::TemplateService
+  include Callable
+
+  class UnknownTemplateTypeError < StandardError; end
+
+  def initialize(current_account, params)
+    @params = params
+    @current_account = current_account
+    @section_id = params.delete('section_id')
+    @section_name = params.delete('section_name')
+    @layout_id = params.delete('layout_id')
+    @layout_name = params.delete('layout_name')
+  end
+
+  attr_reader :current_account, :params, :section_id, :section_name, :layout_id, :layout_name
+
+  class Create < Admin::Api::CMS::TemplateService
+
+    def initialize(current_account, params)
+      super current_account, params
+      @type = params.delete('type')&.to_sym
+    end
+
+    attr_reader :type
+
+    def call
+      collection = { page: current_account.pages,
+                      partial: current_account.partials,
+                      layout: current_account.layouts }[type]
+
+      raise UnknownTemplateTypeError, "Unknown template type '#{type}'" unless type && collection
+
+      template = collection.new(params)
+      attach_section_layout(template)
+      template.save
+
+      template
+    end
+
+    private
+
+    def attach_section_layout(template)
+      return unless type == :page
+
+      template.section ||= find_section
+      template.layout ||= find_layout
+    end
+  end
+
+  class Update < Admin::Api::CMS::TemplateService
+    def initialize(template, current_account, params)
+      super current_account, params
+      @template = template
+      @type = template.class
+    end
+
+    attr_reader :template, :type
+
+    def call
+      attach_section_layout
+      template.update(params)
+
+      template
+    end
+
+    private
+
+    def attach_section_layout
+      return unless type == CMS::Page
+
+      section = find_section
+      template.section = section if section.present?
```

Comment:
That piece of logic is not right.
As the `find_section` always returns something (i.e. the Root section, in case section-related params are missing), an existing section will be overriden to Root in case neither `section_id` or `section_name` are provided, and we don't want that.

Reply:
You're right, I fixed it and also fixed the test, which was wrong.